### ISSUE TITLE
Presence updates hang my game

### ISSUE THREAD

**Issue by ndarilek** [jewlexx/discord-presence#2]
Hey, thanks a bunch for this plugin!

I'm noticing that changing my presence seems to hang my game. I have the following system:

```
fn update_presence(
    app_state: Res<State<AppState>>,
    mut last_state: Local<Option<AppState>>,
    mut activity_state: ResMut<ActivityState>,
    level: Query<&Level, Changed<Level>>,
) {
    let mut changed = false;
    for _ in level.iter() {
        println!("New level");
        changed = true;
    }
    if changed || *last_state != Some(*app_state.current()) {
        println!(
            "App state changed: {:?} != {:?}",
            *last_state,
            Some(*app_state.current())
        );
        match app_state.current() {
            AppState::Loading | AppState::Paused => {}
            AppState::InGame
            | AppState::Exploration
            | AppState::BetweenLives
            | AppState::LevelUp => {
                for level in level.iter() {
                    // activity_state.details = Some(format!("Level {}", **level));
                }
            }
            _ => {
                // activity_state.details = None;
            }
        }
    }
    *last_state = Some(*app_state.current());
}
```

I've confirmed that these conditionals run when they should, and not excessively. As shown, this system doesn't seem to hang the game. However, if I uncomment the line that changes the state details, the game hangs. Likewise, if I uncomment more of the matches so I can set state updates more liberally, the game hangs at each point where an update is sent.

I'm guessing the client isn't async? Wondering if it'd make sense to run the updates in a separate task?

Thanks again.

**Comment by jewlexx** [comment]
Hey sorry or the late reply, I'll take a look in my own environment quickly.

**Comment by jewlexx** [comment]
I did my best to re-create this in my own environment, but it did not introduce any problems for me.

Could you list the following information:

- OS (eg. Windows 10)
- Build target (if not the same as host OS)
- Rust compiler version (eg. rustc 1.60.0 (7737e0b5c 2022-04-04))
- Rust toolchain (eg. nightly)

**Comment by jewlexx** [comment]
Do you get this error per chance (should print in the console where you ran the app):
```
Failed to connect: IoError(Custom { kind: TimedOut, error: "timed out while reading from pipe" })
```

**Comment by ndarilek** [comment]
OS: Windows 10
Rust: 1.60.0 stable
Target: same as build environment

No connection failures. I just experience hangs whenever I uncomment any code that updates my activity.

I'll slap a few printlns in the plugin code in the next day or two and see if I can track down the moment when it hangs.

**Comment by jewlexx** [comment]
Okay, let me know if you make any progress. I'm sorry I can't be of more assistance

**Comment by jewlexx** [comment]
Closing for inactivity. Please re-open or open a new issue if you experience the problem again.

**Comment by ndarilek** [comment]
Sorry for the delays, had some PC issues and needed to get caught up on other things before digging into this.

I've done a bit more digging. I've sprinkled some println!s around the code. It seems like the initial presence set goes through quickly. Subsequent sends hang for a while before responding, though they do eventually go through. The Discord responses don't appear to be errors, they just take a long time to return. Further, there's a pretty large CPU spike. I don't see anything useful in the console other than the messages I've added.

The specific place where the delay seems to happen is here:

```
        println!("Sent");
        let Message { payload, .. } = self.connection_manager.recv()?; // hang
        println!("Received {:?}", payload);
```

I'm not a networking expert and certainly don't know much about interfacing with Discord, but you have a `send_and_receive` function that appears to be interfacing with crossbeam channels, but ultimately it calls `socket.recv()` which, according to its docs, makes no guarantees about whether or not it blocks.

You're also spawning `send_and_receive_loop` on another thread and locking the connection. Is there anything else that needs to hold this lock? I'm wondering if it should be dropped before the `sleep` call to avoid deadlocks, but if nothing else needs it then it wouldn't matter anyway. Adding that didn't make a difference here--I'm just stuck on that `socket.recv` call that makes no guarantees about whether or not it blocks.

Let me know if there's anything else you'd like me to try.

**Comment by ndarilek** [comment]
Er, actually sorry, it's `std::io::Read::read`, not `recv`. Sorry, screen reader user and too many similar-sounding short words.

Anyhow, I've confirmed that the `read` call in `Connection::recv` blocks, at least here. Given it's called in a function managing crossbeam channels, I suspect that's undesirable behavior. :)

**Comment by ndarilek** [comment]
OK, sorry, one more and I'm done for the day. :)

In `WindowsConnection`, you set a read/write timeout of 30 seconds. Curious why these values are so high? Would it be better to make these values lower so the `read` call doesn't block, then silently eat the timeouts in `send_and_receive`?

If I set the value to 1, I get a bunch of timeouts logged. So maybe my Discord client responds slower, but read timeouts on my end are looking like the cause. Maybe presence updates are rate-limited?

Sorry again for all the noise on this issue.

**Comment by jewlexx** [comment]
> Er, actually sorry, it's `std::io::Read::read`, not `recv`. Sorry, screen reader user and too many similar-sounding short words.
> 
> Anyhow, I've confirmed that the `read` call in `Connection::recv` blocks, at least here. Given it's called in a function managing crossbeam channels, I suspect that's undesirable behavior. :)

Good catch :)

(I believe), the recv call was blocking the active thread until something came through, creating your hanging issue.
Could you check-out the `fix/recv` branch and let me know if that version fixes your issue? As I have been unsuccessful in re-creating it on my end.

> In WindowsConnection, you set a read/write timeout of 30 seconds. Curious why these values are so high? Would it be better to make these values lower so the read call doesn't block, then silently eat the timeouts in send_and_receive?

In all honesty, I didn't write most of the base code for the RPC library, I borrowed it from an older library and adapted it to work better with the newer Rust ecosystem, and interact nicely with Bevy. There are a lot of things I am still looking to change in the base library, mostly just a lot of refactoring.

Regardless another good catch, I'll lower those values because we certainly don't want it waiting 30 seconds for a connection that will never arrive.

> Sorry again for all the noise on this issue.

Don't apologize it's helpful to have feedback, realistically I can't fix an issue if nobody talks to me about it so thank you for being active :)

**Comment by ndarilek** [comment]
crates\rpc\src\connection\manager.rs:56 should be `time::Duration`.

It no longer hangs, but I still get timeouts. I suspect what should happen is that the timeouts on read shouldn't be errors, they're just a condition of the read blocking. Also, should 0 bytes received be classified as connection loss, or just an empty read?

Thanks for digging into this.

**Comment by jewlexx** [comment]
> I suspect what should happen is that the timeouts on read shouldn't be errors, they're just a condition of the read blocking.

Yes and no.

True that they shouldn't exit the program (which I believe they don't), but they are a case where we have sent a message without a response which isn't supposed to happen, I don't believe.

What I will do for now, is change the log to debug rather than error so that it won't show up in production builds and shouldn't be as annoying. I'll keep looking though and try and work out the structure.

**Comment by ndarilek** [comment]
To be clear, a response eventually arrives, so it isn't strictly 
accurate that you're sending without one. It just responds late.

 > Discord_UpdatePresence() has a rate limit of one update per 15 
seconds. -- https://discord.com/developers/docs/rich-presence/how-to

If I'm not mistaken, you send one when the app starts. I send one when 
my menu opens. Assuming the rate limit manifests by hanging the RPC 
connection until it can be processed, so unblocking becomes the cue that 
the next limit can be sent, it makes sense that it'd take about 30 
seconds or so for my menu to appear. Then there's another delay when I 
start the game and send "Level 1". So in short, I don't think these are 
errors at all. Maybe OS-level differences in how the RPC socket works. 
That may be how the rate limit is communicated.

**Comment by jewlexx** [comment]
Apologies for the late response.

Could you check the updated version of the library and see if that fixes any of the issues you were having?

**Comment by ndarilek** [comment]
With 0.3.2, the game doesn't hang, but I get a long stream of timeout 
errors at about the rate of 1/15 seconds.Message ID: 
***@***.***>

**Comment by ndarilek** [comment]
https://github.com/discordjs/RPC/issues/14#issuecomment-818805016 Not sure if that particular field is being included, but it seems like removing it helped in some instances.

**Comment by jewlexx** [comment]
> Not sure if that particular field is being included

By default, it shouldn't be. If the developer sets `show_time` to true, then it does, in order to display the time in Discord but otherwise it doesn't.

**Comment by jewlexx** [comment]
If you delete Cargo.lock and re-install dependencies, it should update the base library version to 0.5.5 which should fix the logs

**Comment by ndarilek** [comment]
Sorry for the delay. Doesn't appear as if 0.5.5 is released.

**Comment by jewlexx** [comment]
My bad. Should be published now

**Comment by jewlexx** [comment]
@ndarilek Any updates?

**Comment by ndarilek** [comment]
Sorry, looks like I replied via email but the comment isn't listed here. On July 8 I wrote:

Ugh, still just reports multiple timeouts. 

Thanks for looking into this, not sure what else to recommend.

**Comment by jewlexx** [comment]
That's unfortunate. If you could try it again with v0.5.6? It fixed the issue with #12 which may be unrelated but worth a shot I guess.

**Comment by ndarilek** [comment]
Same log message, but now the game hangs again and doesn't launch. :( Maybe it'd time out eventually but I didn't wait.

**Comment by jewlexx** [comment]
Hey sorry for the late response. I think I am finally able rto recreate the issue on my local environment, so I can finally properly get to fixing it. I will update you asap.

**Comment by jewlexx** [comment]
Hey, I think I may have finally fixed the issue in #20. If you could test it for me that would be great thanks :)

**Comment by ndarilek** [comment]
It does seem to hang a *lot* less, though there are occasional spots where the game loop still gets stuck which may correspond to places where updates happen. I'm trying to investigate whether or not I'm being rate-limited or similar, since it's always in my menus and not in the same place. But yes, I now have the plugin enabled because at the very least I can play the game and not have it hang.

I have the following in Cargo.toml, hopefully this is the right dependency combination:

```
bevy-discord-presence = "0.3"
[patch.crates-io]
discord-presence = { git = "https://github.com/jewlexx/discord-presence" }
```

Going to do a bit more debugging of the remaining hangs, and will report back what I find.

**Comment by jewlexx** [comment]
> I have the following in Cargo.toml, hopefully this is the right dependency combination:
> 
> ```
> bevy-discord-presence = "0.3"
> [patch.crates-io]
> discord-presence = { git = "https://github.com/jewlexx/discord-presence" }
> ```

That's perfect. I have published the blocking patch but until it is fixed altogether, if you could stay on the github branch that would be great thanks.

> It does seem to hang a lot less, though there are occasional spots where the game loop still gets stuck which may correspond to places where updates happen.

Glad to hear it is hanging less, I will dig through the updates code and I am considering working on a built in rate limiter, which queues executions until the rate limit is refreshed, as currently it is very easy for developers to hit the rate limit by accident.

**Comment by ndarilek** [comment]
Ugh, still just reports multiple timeouts.

Thanks for looking into this, not sure what else to recommend.

**Comment by jewlexx** [comment]
Hey @ndarilek I am currently working on a patch that will use websockets rather than native sockets, which might fix the issues you were having. It's not close to being done but you can track progress here: #28

**Comment by ndarilek** [comment]
Finally getting time to circle back to this.

Were the websockets changes ever merged?

Re-enabling the released version of the plugin in my game now gives this message in addition to the others I've seen:

2023-01-09T14:31:59.151109Z ERROR bevy_discord_presence: Failed to set presence: Connection has not been started                                                          

And then I get the usual hang.

Thanks.

**Comment by jewlexx** [comment]
No sorry websockets were never merged, I never finished working on it. I've spent a long time away from this library.
The `Connection has not been started` is related to #37.
I will continue working on websockets ASAP.

**Comment by jewlexx** [comment]
This issue hasn't had any updates for years now. I'm gonna close it, if you still have issues please feel free to reopen.
